Add handling for comics with Early Access

Ruthalas opened the pull request:
Skips early access blocked strips instead of crashing.
Uses two try/except pairs, one for page parsing, and one for page downloading.

It's not exceptionally elegant, but it works!

This pull would resolve Issue #21

Ruthalas committed:
Add handling for comics with Early Access

Skips early access blocked strips instead of crashing. Uses two try/except pairs, one for page parsing, and one for page downloading.

TilCreator commented:
That's just a catch all which is one of the bad practices ever, it especially makes debugging a pain.
My tip is to get the URL with requests and check for a 404 before parsing with pyquery.
Example: https://github.com/TilCreator/Tapas-Comic-Downloader/blob/cb42561e0627fca2c0045b519968e1bf951fa64d/tapas-dl.py#L74-L80

Ruthalas commented:
That is very true. I am happy to rewrite in that fashion.

Is there a nice way to then handle the missing img slots when the next loop goes through to download them? To be honest, the way you've constructed the loops with enumerate is a bit beyond my ken. I haven't quite grasped it.

TilCreator commented:
`enumerate` is just so I can get how far the for loop is through the list and print the progress.

TilCreator commented:
My idea would be to truncate the `data` list.

So `data = data[:pageCount]` after the first for loop you changed.
But that also requires the first for loop to break after an 404 of course.

TilCreator commented:
Maybe [`for - else`](https://book.pythontips.com/en/latest/for_-_else.html) could be used there.

Ruthalas commented:
I've adjusted the initial page parse to check for a 404 error as you described.
I think truncating the dict blindly (ie: truncating it to the length of noted pages) is crude. If the 404's are anywhere but grouped at the end of the list, we'll be truncating the wrong items. Breaking the loop after the first 404 seems to cause the same issue.

I haven't yet found an elegant way to check for an out-of-bounds when going through the resulting data. I am continuing to look at that.

Edit: Perhaps adding single dummy image listing, with a name like '404', and simply ignoring those entries when downloading the images?

Ruthalas committed:
Changed "Early Access" handling

Changed early access handling not to rely on try/except, but rather to check for 404 errors on page access. It places a dummy entry in the img download list, which is checked for when downloading images,

Ruthalas committed:
Changed user feedback on image download

Changed the 'downloaded x images' to read, ' downlaoded y of x images' where y is the count of images we actually downloaded, and x is still the total number we found as prospective downloads.

Ruthalas committed:
Added comments, cleaned up

Ruthalas committed:
Improved comments and readability

Ruthalas committed:
Re-enabled book check

I had this disabled for testing

Ruthalas commented:
The current revision incorporates your suggestion to check for non-200 page request responses to determine page availability.
When if finds a page that yields another response, it informs the user and adds a 'pageUnavailable' entry to the img list.

![image](https://user-images.githubusercontent.com/27835519/81508014-ef21d680-92f0-11ea-9000-bafb0dfca62d.png)

When looping over the dictionary to download images, each page's first entry is checked, and if it matches 'pageUnavailable', that entry is skipped (notifying the user).

![image](https://user-images.githubusercontent.com/27835519/81508028-f9dc6b80-92f0-11ea-9d35-a73ac4784c0b.png)

Finally the 'Downloaded x images' message was modified to show 'Downloaded y of x images' where y is the count of images we actually downloaded, and x is still the total number we found as prospective downloads.

![image](https://user-images.githubusercontent.com/27835519/81508075-42942480-92f1-11ea-872d-fe80eee5c306.png)

Let me know what you think of that solution!
Thanks for your time.

TilCreator commented:
I approve of the changes, ready for merge?

Ruthalas commented:
Yes, go for it!

TilCreator commented:
Thank you!
